pkcs11/signer.go

```diff
@@ -168,7 +167,12 @@ func (s *signer) SignX509Cert(ctx context.Context, cert *x509.Certificate, keyId
 		return nil, errors.New("client request timed out, skip signing X509 cert")
 	}
 	defer pool.put(signer)
-	cert.SignatureAlgorithm = getSignatureAlgorithm(signer.signAlgorithm())
+	// Validate the cert request to ensure it matches the keyType and also the HSM supports the signature algo.
+	if val := isValidCertRequest(cert, signer.publicKeyAlgorithm(), signer.signAlgorithm()); !val {
+		log.Printf("SignX509Cert: unsupported signature algo %d, supported algo %d", cert.SignatureAlgorithm, signer.signAlgorithm())
+		// Not a valid signature algorithm. Overwrite it with what the configured keyType supports.
+		cert.SignatureAlgorithm = x509cert.GetSignatureAlgorithm(signer.signAlgorithm())
```

Comment:
Why do we only overwrite `SignatureAlgorithm` given that `isValidCertRequest` checks both `PublicKeyAlgorithm` and `SignatureAlgorithm`?

Besides, I'm not sure what we can gain from adding `isValidCertRequest`. The logs are not associated with a request id, so we would only know that some cert requests contain invalid values. We may just keep the original behavior (i.e. simply overwrite what's in the CSR with what we supports) or bail out by returning an error.